File: backend/assemblyai.py
from pprint import pprint
from time import sleep
import requests

API_KEY_ASSEMBLYAI = 'cd73c652ca7a4e6282e588cc8d5cee6d'

upload_endpoint = 'https://api.assemblyai.com/v2/upload'
transcript_endpoint = 'https://api.assemblyai.com/v2/transcript'

headers_auth_only = {
  'authorization': API_KEY_ASSEMBLYAI
}
headers = {
  'authorization': API_KEY_ASSEMBLYAI,
  'content-type': 'application/json'
}

BATCH_SIZE = 5 * 1024 * 1024 # 5 MB

def upload_file(filename):
  def read_file(filename):
    with open(filename, 'rb') as f:
      while True:
        data = f.read(BATCH_SIZE)
        if not data: return
        yield data

  upload_response = requests.post(
    upload_endpoint,
    headers = headers_auth_only,
    data = read_file(filename)
  )

  pprint(upload_response.json())
  return upload_response.json()['upload_url']

def transcribe(audio_url):
  transcript_request = {
    'audio_url': audio_url,
    'auto_chapters': True,
    'auto_highlights': True
  }

  transcript_response = requests.post(
    transcript_endpoint,
    json = transcript_request,
    headers = headers
  )

  pprint(transcript_response.json())
  return transcript_response.json()['id']

def poll(transcript_id):
  polling_endpoint = f'{transcript_endpoint}/{transcript_id}'
  polling_response = requests.get(
    polling_endpoint,
    headers = headers
  )

  if polling_response.json()['status'] == 'completed':
    return {
      'transcript': polling_response.json()['text'],
      'chapters': polling_response.json()['chapters'],
      'highlights': polling_response.json()['auto_highlights_result']
    }
  return False

def pipeline(filename):
  audio_url = upload_file(filename)
  transcribe_id = transcribe(audio_url)
  while True:
    result = poll(transcribe_id)
    if result: return result
    print('Still processing! Trying again after 60 seconds...')
    sleep(60)

pipeline('./test_short.m4a')

File: frontend/src/assembly_ai.js
import axios from "axios"

const API_KEY_ASSEMBLYAI = 'cd73c652ca7a4e6282e588cc8d5cee6d'

const upload_file = async (file) => {
  const assembly = axios.create({
    baseURL: "https://api.assemblyai.com/v2",
    headers: {
      authorization: API_KEY_ASSEMBLYAI,
    },
  });
  const result = await assembly.post("/upload", file)
  return result ? result.data.upload_url : ''
}

const transcribe = async (audio_url) => {
  const assembly = axios.create({
    baseURL: "https://api.assemblyai.com/v2",
    headers: {
        authorization: API_KEY_ASSEMBLYAI,
        "content-type": "application/json",
    },
  });
  const result = await assembly.post("/transcript", { audio_url, auto_chapters: true })
  return result ? result.data.id : ''
}

const poll = async (transcript_id) => {
  const assembly = axios.create({
    baseURL: "https://api.assemblyai.com/v2",
    headers: {
        authorization: API_KEY_ASSEMBLYAI,
        "content-type": "application/json",
    },
  });
  const result = await assembly.get(`/transcript/${transcript_id}`)
  return result ? result.data : null
}

export const pipeline = async (file) => {
  const audio_url = await upload_file(file)
  if (!audio_url) return null
  const transcript_id = await transcribe(audio_url)
  if (transcript_id) {
    while (true) {
      const result = await poll(transcript_id)
      if (result && result.status === 'completed') {
        return {
            transcript: result.text,
            chapters: result.chapters,
            highlights: result.auto_highlights_result
          }
      } else if (result && result.status === 'error') {
        return null
      }
      console.log('Still processing! Trying again after 60 seconds...')
      await new Promise((resolve) => setTimeout(resolve, 60000))
    }
  }
}